V3/remove this throw call transaction h

gberkes opened the pull request:
Introducing the use of assertions to address throw; calls that lack try-catch blocks. Upon examining the caller code that utilized methods containing the questioned throw; calls, it became clear that, in the current state of development, there are no scenarios where execution could reach these throw; calls. However, we cannot guarantee this for future development. For instance, if someone attempts to use getCurrentMarker() without first verifying isInsideAMarker(), ModSecurity would encounter the throw; and terminate. The issue with the other throw; call is similar in that it is, fortunately, unreachable at the moment. However, it differs because this throw; is intended to handle a case that has not yet been developed.

Fortunately, I found an article titled "Effective Use of Assertions in C++" by Mike A. Martin in (ACM SIGPLAN Language Tips, page 3), which offers a neat way to handle such cases, specifically regarding argument validation and unreachable code. [Link to the article](https://dl.acm.org/doi/pdf/10.1145/240964.240969).

Following the guidance from this article, I addressed the issues and also included modifications to enrich assert error messages. Furthermore, I updated configure.ac to maintain the usual build procedure and modified README.md to introduce the new configure flag.

Commit:
Refactor: Ensure safe error handling by removing isolated `throw;` statements.

- SonarCloud analysis identified standalone `throw;` calls without accompanying `try-catch` blocks, used inconsistently as placeholders or for premature termination under specific conditions.
- Removed these `throw;` instances to prevent potential runtime issues in future development phases, where such configurations might inadvertently be created.
- Introduced `assert` statements as a more appropriate mechanism for asserting preconditions in the affected class member functions, ensuring clearer intent and safer code behavior during development.

Commit:
Build System: Introduce Configurable Assertion Handling

Implemented a new configuration option --enable-assertions=[yes|no] within config.ac, enabling controlled inclusion of -DNDEBUG in CPPFLAGS. The default setting suppresses assertions (by adding -DNDEBUG to CPPFLAGS), preserving the original behavior. This enhancement allows for the optional enabling of assertions during development or debugging by setting --enable-assertions=yes, thereby excluding -DNDEBUG from CPPFLAGS.

Commit:
Making your assert statements more descriptive

Source: https://www.learncpp.com/cpp-tutorial/assert-and-static_assert/

gberkes committed:
Merge branch 'owasp-modsecurity:v3/master' into v3/remove_this_throw_call_transaction_h

gberkes committed:
Merge branch 'owasp-modsecurity:v3/master' into v3/remove_this_throw_call_transaction_h

Commit:
Update README.md to recommend --enable-assertions=yes for debugging sessions.

Commit:
Fix null pointer dereference error in 'src/rule_with_actions.cc'

Resolved a null pointer dereference error identified by `make check-static` due to an outdated suppression in 'test/cppcheck_suppressions.txt'. The suppression at line 57 was incorrectly referencing 'src/rule_with_actions.cc:244'. Updated this to 'src/rule_with_actions.cc:243' to align with the current codebase.

Additionally, removed suppressions for 'rethrowNoCurrentException' as identified by SonarCloud. This was possible due to the removal of standalone 'throw;' statements in the code.

Commit:
Enable assertions in QA testing environment to improve error detection.

airween commented on headers/modsecurity/transaction.h:
```diff
@@ -307,11 +308,8 @@ class TransactionSecMarkerManagement {
-        if (m_marker) {
-            return m_marker;
-        } else {
-            throw;
```
SonarCloud report is [here](https://sonarcloud.io/project/issues?resolved=false&types=BUG&id=owasp-modsecurity_ModSecurity&open=AY1CfJimrsSpWCKX0vrq)

airween commented on src/rule_with_actions.cc:
```diff
@@ -121,10 +122,8 @@ RuleWithActions::RuleWithActions(
                 delete a;
-                std::cout << "General failure, action: " << a->m_name;
-                std::cout << " has an unknown type." << std::endl;
-                throw;
```
SonarCloud report is [here](https://sonarcloud.io/project/issues?resolved=false&types=BUG&id=owasp-modsecurity_ModSecurity&open=AY1CfKAnrsSpWCKX0waI)

airween commented:
Just for the record: this PR fixes two SonarCloud issues in files:

* [headers/modsecurity/transaction.h](https://sonarcloud.io/project/issues?resolved=false&types=BUG&id=owasp-modsecurity_ModSecurity&open=AY1CfJimrsSpWCKX0vrq)
* [src/rule_with_actions.cc](https://sonarcloud.io/project/issues?resolved=false&types=BUG&id=owasp-modsecurity_ModSecurity&open=AY1CfKAnrsSpWCKX0waI)

I also added these references as conversation.

airween commented on README.md:
```diff
@@ -231,7 +231,7 @@ CFLAGS to disable the compilation optimization parameters:
 $ export CFLAGS="-g -O0"
 $ ./build.sh
-$ ./configure
+$ ./configure --enable-assertions=yes
```
Please explain the behavior of this new option. First it's enough to place that in a single comment.

Later we should consider to create a separated documentation / extend the existing one with all options.

gberkes replied:
Done: https://github.com/owasp-modsecurity/ModSecurity/pull/3104/commits/2c8684eeb0e99b25a4d488d802910cba927153bb

MirkoDziadzka commented on headers/modsecurity/transaction.h:
```diff
@@ -323,7 +321,7 @@ class TransactionSecMarkerManagement {
 
  private:
-    std::shared_ptr<std::string> m_marker;
+    std::shared_ptr<std::string> m_marker = nullptr;
```
I think C++ gurantees this. `std::shared_ptr` is a class and therefore has a constructor and initialize itself.

gberkes replied:
Yes. You're right. Fixed: https://github.com/owasp-modsecurity/ModSecurity/pull/3104/commits/cee3965f689ea52004d0e8b160054e227e6c5593

MirkoDziadzka commented on src/rule_with_actions.cc:
```diff
@@ -121,10 +122,8 @@ RuleWithActions::RuleWithActions(
             } else {
+                assert(false && "The handling of RunTimeBeforeMatchAttemptKind has not been implemented yet.");
                 delete a;
-                std::cout << "General failure, action: " << a->m_name;
-                std::cout << " has an unknown type." << std::endl;
-                throw;
```
What we are doing here is a switch over `a->action_kind` withe the 2 cases `Action::ConfigurationKind` and `Action::RunTimeOnlyIfMatchKind`.

This `else` would be the default case of the switch, right.

Would it be better to replace the if/else with a switch and ask the compiler of our switch case is complete (e.g. all cases are implemented)

----

independent from this remark: Is the else case a "can never happen" or is it a "user configured invalid modsec code". In the second case, we change the semantic which could be a problem.

gberkes replied:
The refactoring of the if-then chain to a switch statement is complete. We reviewed the code together with @airween  and couldn't find a way to force execution into the else/default case.

gberkes committed:
Merge branch 'owasp-modsecurity:v3/master' into v3/remove_this_throw_call_transaction_h

Commit:
Explicitly use -UNDEBUG when --enable-assertions=yes is specified.

Commit:
Document importance of reviewing cppcheck_suppressions.txt during development and usage of assertions.

Commit:
Remove unnecessary nullptr initialization from m_marker.

Commit:
Refactor action_kind processing to use switch() instead of if-else chains; add assertion in default case.

Commit:
Delete code breaking cppcheck.

Commit:
Align suppression line number.

Commit:
Refactor to use cppcheck-friendly conditional inclusion of break/assert() statements.

Commit:
Fix SonarCloud issue: Make this variable a const reference.

https://sonarcloud.io/project/issues?resolved=false&pullRequest=3104&id=owasp-modsecurity_ModSecurity&open=AY8Vpgy4f6U6E7VKL4Cn

airween commented:
Please pull the modifications from #3134, and apply this PR. Or you can resolve the conflicts [here](https://github.com/owasp-modsecurity/ModSecurity/pull/3104/conflicts).

huberts90 commented:
@gberkes
Could you please address requested change to complete this issue?
